### ISSUE TITLE
Scope out how we might move to a single source of truth data model

### ISSUE THREAD

**Issue by danswick** [GSA-TTS/FAC#4508]
Tweaking this ticket from its original form.

**squad**: @gsa-jrothacker @sambodeme @asteel-gsa @danswick 

```[tasklist]
### Outcomes:
- [ ] Sketch out each stage data flows through and what value we get from each structure and/or transformation for each
- [ ] Identify opportunities to simplify 
- [ ] Walk through a scenario in which we might drop every version of the data except one. What would be lost? What would we need to do to ensure no real data loss occurs?
``` 

<details>
<summary>Prior discussion from dev huddle</summary>
During today's dev huddle, we discussed some of the challenges around DB standup/teardown, running backups without interrupting the running app, and the white whale :whale: of re-disseminating submissions on-demand.

One of my questions was, essentially, "do we still need the dissemination tables?" We need them to support basic search as well as API's table-based endpoints, but could what if we went back to a single search (basically the current "advanced")? And what if we reconsidered the API in a way that didn't require table-by-table endpoints mirroring Census' old downloads?

This spike is essentially a place for us to 1) discuss what an ideal model would look like if we put all the current system's constraints out of mind and 2) run an experiment to see what it would look like to treat the `intake` table as a single source of truth that gets rendered directly into the schema/model we need to support advanced search, eliminating the intermediary dissemination table completely. 

Everything here is up for discussion. I'm capturing this as a starting point.

Some needs/wants we might consider:
* How might we handle re-disseminating individual records on-demand?
* If we use the intake table as a single source of truth, how might that impact our backup strategy? 
* Where might a read replica DB make sense in any of these hypothetical scenarios?
* If we move away from the current table-by-table API endpoints, how might we gracefully deprecate the prior iteration? 

Some possible goals:
* Protect the rest of the app from huge spikes in API traffic.
* Create a clear path for re-dissemination.
* Improve long-term maintainability.
* Streamline deploy, setup, tear-down, and backups.
* Identify opportunities to "pre-render" or otherwise pre-process data to support our users' needs (better excel exports, bulk downloads, search, API perf, etc). 

cc all the folks who chimed in during discussion: @asteel-gsa @rnovak338 @sambodeme @gsa-jrothacker @gsa-suk @jperson1 
</details>

**Comment by rnovak338** [comment]
Some of my thoughts from our discussion.

## Ideal case

We take the "source of truth" burden off of the dissemination tables and put it on the intake tables. With this change, any data that is generated/formatted based on intake data **should** be disposable. Meaning, we can tear down / rebuild / modify this data however we need to for the end user's display. It also means we could feel more comfortable about cases like resubmitting a report. Lastly, we are also removing data redundancy between the dissemination tables, and whichever place Advanced Search/API would point to.

## How do we make the ideal case work?

### Dissemination

- Don't we need to have an "official" FAC acceptance date to prove the user has submitted data within their deadline? (And to show in cases where it may need resubmission, that they were on time in the first place)
- In the ideal case, wouldn't this mean we have to put this date into intake somehow, as any other place would no longer make it the "source of truth", and would challenge our ability to dispose of and rebuild the data in the ideal scenario.

### Performance

Referencing Matt's original insights here: (https://github.com/GSA-TTS/FAC/issues/4285#issuecomment-2391549639).
With his PR, we "could" still respect the above and cut down on dissemination / basic search. His changes point Advanced Search and API to a series of tables in DB-2 (which are based off of dissemination, NOT intake currently).

I really like Hassan's comment in our slack conversation so I will include it here as well.

---
One of the main issues Matt's draft PR aims to address is performance and he tackled it through:
1. Creating a read-replica database to ensure that general read queries do not impact the intake process, which is critical during peak seasons such as March.
6. Simplifying data access control by separating the public records schema from the tribal records schema.
7. Reducing restoration time by eliminating materialized views (MVs) and relying on actual tables instead.
8. Improving overall query response times for heavy queries by implementing the above changes, particularly through schema updates.

If we only create a read replica of the current database and route API and potentially search queries to it, we will still need to address items 2 through 4 (or items 3 and 4 if item 2 is not a significant concern).

---

**Comment by danswick** [comment]
Thanks, @rnovak338!  

>If we only create a read replica of the current database and route API and potentially search queries to it...

That is not what this ticket proposes. I'm suggesting we turn intake entries _directly_ into entries in the DB/tables that power the API and search interfaces, bypassing the intermediate dissemination tables we use today.

The other bits:

* The acceptance date info is already saved to the SF-SAC record. The disseminated record is just a copy of that info. It's _also_ a more brittle way to track those dates as we've seen with the "stuck SACs". 
* The performance bottleneck improvements have nothing to do with the dissemination tables and potentially commit us to more complexity and debugging difficulty down the road. 
  * There's nothing preventing us from rendering intake tables directly into a read-replica DB with whatever schema we need to support both the API and advanced search. Basic search would go away forever since it was really just a workaround during our first busy season. Similar for the MV point: piping submissions directly into tables that meet our needs eliminates the need for a MV *and* multiple DBs for different kinds of search/API activities. Further, if we can create a pipeline in which _any record can be re-disseminated at any time without drama_, restoration isn't as much of an issue.
  * If the above is possible, performance improvements should be the same as the other approach. 
* I don't agree that access control is simplified by replicating data across two databases. Most tribal report data is public. Only some elements can be suppressed. 

Today, we have _two sources of truth_, which has proven to be an absolute nightmare for maintenance and troubleshooting. If the dissemination and intake tables get out of sync, things just start breaking. It's taken _six months_ to work through the final stages of historical data migration specifically because of this brittle relationship between intake and dissemination. We're still completely stuck trying to figure out a suitable pathway to re-disseminating reports and the only way forward seems to be _even more duplication_. I'm _very concerned_ about adding not only another DB, but also investing in more complex setup/teardown and replication processes to support all of the above. The more we invest in performance tuning the current system, the harder it will be to unwind later on, so now's the time to investigate other approaches.

To summarize,  **the dissemination tables aren't suitable for _actually disseminating data_ and we need to seriously consider replacing them with something else**.

**Comment by danswick** [comment]
This kind of stuff is happening elsewhere.
